include/bst.hpp

```diff
@@ -0,0 +1,45 @@
+#ifndef __BINARY_SEARCH_TREE__
+#define __BINARY_SEARCH_TREE__
+
+template <class T>
+class BNode {
+public:
+    const T& operator*() const;
+    T& operator*();
+};
+
+template <typename T>
+class BinarySearchTree {
+private:
+    
+public:
+    typedef BNode* iterator;
+
+    BinarySearchTree();
+    ~BinarySearchTree();
+
+    BinarySearchTree(const BinarySearchTree &other);
+    BinarySearchTree &operator=(const BinarySearchTree &other);
+
+    void insert(const T &value);
```

Comment:
It would be good if you can accept another template param type instead of class type for flexibility:
```cpp
BinarySearchTree<std::string> bst;
bst.insert("Hello"); // pass `const char*` instead of `std::string`
```